File: src/main/java/chatbot/enums/Command.java
package chatbot.enums;

import java.util.HashMap;
import java.util.Set;

public enum Command {
  NAMMERS(false, true, true, true, false),
  NAMPING(false, true, true, true, false),
  NAMES(false, true, true, true, false),
  NAMREFRESH(false, false, false, false, false),
  NAMCOMMANDS(false, true, true, true, false),
  NAMCHOOSE(false, true, true, true, false),
  NAM(false, true, true, true, false),
  LASTMESSAGE(false, false, true, true, true),
  LM(false, false, true, true, true),
  FIRSTMESSAGE(false, true, true, true, true),
  FM(false, true, true, true, true),
  LOG(true, true, true, true, false),
  LOGS(true, true, true, true, false),
  RQ(false, true, true, true, true),
  RS(false, true, false, true, true),
  ADDDISABLED(false, true, false, true, false),
  REMDISABLED(false, true, false, true, false),
  FS(false, false, false, false, false),
  SEARCH(false, true, true, true, false),
  SEARCHUSER(false, true, true, true, false),
  ADDALT(false, false, false, true, false),
  NAMBAN(false, false, false, true, false),
  SC(false, false, false, true, false),
  LASTSEEN(true, false, true, true, false),
  LS(true, false, true, true, false),
  MCOUNT(true, true, true, true, false)
  ;


  private final boolean onlineAllowed;
  private final boolean selfAllowed;
  private final boolean othersAllowed;
  private final boolean modsAllowed;
  private final boolean canOptOut;

  /**
   * @param onlineAllowed command is allowed to be used in onlinechat.
   * @param selfAllowed   command can be used with own name as argument.
   * @param othersAllowed command can be used with others name as argument.
   * @param modsAllowed   command can be used by mods.
   */
  Command(boolean onlineAllowed, boolean selfAllowed, boolean othersAllowed, boolean modsAllowed,
      boolean canOptOut) {
    this.onlineAllowed = onlineAllowed;
    this.selfAllowed = selfAllowed;
    this.othersAllowed = othersAllowed;
    this.modsAllowed = modsAllowed;
    this.canOptOut = canOptOut;
  }

  public boolean isOnlineAllowed() {
    return onlineAllowed;
  }

  public boolean isSelfAllowed() {
    return selfAllowed;
  }

  public boolean isOthersAllowed() {
    return othersAllowed;
  }

  public boolean isModsAllowed() {
    return modsAllowed;
  }

  public boolean isOptedOut(String username, Set<String> optOutList) {
    return canOptOut && optOutList.contains(username);
  }

  @Override
  public String toString() {
    return super.toString().toLowerCase();
  }

  public Boolean isUserCommandSpecified(HashMap<String, Boolean> userPermissionMap) {
    return userPermissionMap.getOrDefault(toString(), null);
  }
}

File: src/main/java/chatbot/enums/Command.kt
package chatbot.enums

import chatbot.enums.Command.Feature.*
import java.util.*

public enum class Command {
    NAMMERS(NO_ARGS),
    NAMPING(NO_ARGS),

    NAMES(ONLINE, SELF, OTHERS),

    NAMREFRESH(ADMIN_ONLY),
    NAMCOMMANDS(ADMIN_ONLY),

    NAMCHOOSE(NO_ARGS),

    NAM(SELF, OTHERS),

    LASTMESSAGE(OTHERS, OPT_OUT),
    LM(LASTMESSAGE),

    FIRSTMESSAGE(SELF, OTHERS, OPT_OUT),
    FM(FIRSTMESSAGE),

    LOG(ONLINE, SELF, OTHERS),
    LOGS(LOG),

    RQ(SELF, OTHERS, OPT_OUT),
    RS(SELF, OPT_OUT),

    ADDDISABLED(SELF),
    REMDISABLED(SELF),

    FS,

    SEARCH(NO_ARGS),
    SEARCHUSER(SELF, OTHERS),

    ADDALT,
    NAMBAN,

    SC(ADMIN_ONLY),

    LASTSEEN(ONLINE, OTHERS),
    LS(LASTSEEN),

    MCOUNT(ONLINE, SELF, OTHERS),
    ;


    val isOnlineAllowed: Boolean
    val isSelfAllowed: Boolean
    val isOthersAllowed: Boolean
    val isAdminOnly: Boolean
    val canOptOut: Boolean
    val isNoArgs: Boolean

    constructor(vararg features: Feature) {
        var featureSet: EnumSet<Feature> = EnumSet.noneOf(Feature::class.java)

        if (!features.isEmpty()) {
            featureSet = EnumSet.copyOf<Feature>(listOf<Feature>(*features))
        }

        this.isOnlineAllowed = featureSet.contains(ONLINE)
        this.isSelfAllowed = featureSet.contains(SELF)
        this.isOthersAllowed = featureSet.contains(OTHERS)
        this.isAdminOnly = featureSet.contains(ADMIN_ONLY)
        this.canOptOut = featureSet.contains(OPT_OUT)
        this.isNoArgs = featureSet.contains(NO_ARGS)
    }

    constructor(alias: Command) {
        this.isOnlineAllowed = alias.isOnlineAllowed
        this.isSelfAllowed = alias.isSelfAllowed
        this.isOthersAllowed = alias.isOthersAllowed
        this.isAdminOnly = alias.isAdminOnly
        this.canOptOut = alias.canOptOut
        this.isNoArgs = alias.isNoArgs
    }

    fun isOptedOut(username: String, optOutList: MutableSet<String>): Boolean {
        return canOptOut && optOutList.contains(username)
    }

    fun isUserCommandSpecified(userPermissionMap: MutableMap<String, Boolean>): Boolean? {
        return userPermissionMap.getOrDefault(toString().lowercase(), null)
    }

    internal enum class Feature {
        ONLINE,
        SELF,
        OTHERS,
        ADMIN_ONLY,
        OPT_OUT,
        NO_ARGS,
    }
}
